# Data cleaning Calculation Transformation on PAP data - Aug 30th

In [3]:
import pandas as pd

pap_data = pd.read_csv('C:/Users/anast/pap_location_data.csv')

In [5]:
pap_data.head()

Unnamed: 0           Type_appart Surface m²  Price € Localisation  \
0           0  meublee appartement         42m     1280    Paris 14E   
1           1  meublee appartement         53m     1533       Paris    
2           2  meublee appartement         38m     1400    Paris 13E   
3           3  meublee appartement         33m     1130    Paris 18E   
4           4  meublee appartement         75m     2350     Paris 9E   

              Date Post                                        Description  
0  /  02 septembre 2019  Métro Alésia, proche transports et bus.\n\r\nD...  
1       /  30 août 2019  Daumesnil. Haussmannien, au calme, 3 pièces, 5...  
2       /  30 août 2019  Métro place d'Italie ou Tolbiac.\r\n2 pièces m...  
3       /  30 août 2019  2 pièces de 32,5m2 en parfait état entre Lamar...  
4       /  30 août 2019  Appartement meublé 75,5 m², entièrement rénové...

In [6]:
pap_data.drop(columns=['Unnamed: 0'], inplace=True)

In [7]:
pap_data.head()

Type_appart Surface m²  Price € Localisation  \
0  meublee appartement         42m     1280    Paris 14E   
1  meublee appartement         53m     1533       Paris    
2  meublee appartement         38m     1400    Paris 13E   
3  meublee appartement         33m     1130    Paris 18E   
4  meublee appartement         75m     2350     Paris 9E   

              Date Post                                        Description  
0  /  02 septembre 2019  Métro Alésia, proche transports et bus.\n\r\nD...  
1       /  30 août 2019  Daumesnil. Haussmannien, au calme, 3 pièces, 5...  
2       /  30 août 2019  Métro place d'Italie ou Tolbiac.\r\n2 pièces m...  
3       /  30 août 2019  2 pièces de 32,5m2 en parfait état entre Lamar...  
4       /  30 août 2019  Appartement meublé 75,5 m², entièrement rénové...

In [8]:
pap_data.shape

(200, 6)

In [9]:
pap_data.isnull().sum()

Type_appart     0
Surface m²      0
Price €         0
Localisation    0
Date Post       0
Description     0
dtype: int64

In [10]:
pap_data.Localisation.unique()

array(['Paris 14E', 'Paris ', 'Paris 13E', 'Paris 18E', 'Paris 9E',
       'Paris 11E', 'Paris 17E', 'Paris 2E', 'Paris 10E', 'Paris 16E',
       'Paris 3E', 'Paris 5E', 'Paris 15E', 'Paris 19E', 'Paris 20E',
       'Paris 12E', 'Paris 6E', 'Paris 7E', 'Paris 1Er', 'Paris 8E',
       'Paris 4E', 'Paris 11'], dtype=object)

In [11]:
pap_data[pap_data.Localisation =='Paris ']

Type_appart Surface m²  Price € Localisation        Date Post  \
1    meublee appartement         53m     1533       Paris   /  30 août 2019   
6    meublee appartement         71m     1964       Paris   /  30 août 2019   
11   meublee appartement         53m     1533       Paris   /  30 août 2019   
16   meublee appartement         71m     1964       Paris   /  30 août 2019   
28   meublee appartement         74m     2370       Paris   /  30 août 2019   
42   meublee appartement         46m     1900       Paris   /  30 août 2019   
64   meublee appartement         38m     1100       Paris   /  30 août 2019   
82        meublee studio          9m      450       Paris   /  30 août 2019   
90   meublee appartement         25m      950       Paris   /  30 août 2019   
92        meublee studio          9m      420       Paris   /  30 août 2019   
96        meublee studio          9m      520       Paris   /  30 août 2019   
99   meublee appartement         24m      870       Paris   /  30 août 2019   
104       meublee studio         18m      700       Paris   /  30 août 2019   
105  meublee appartement         37m     1395       Paris   /  30 août 2019   
111       meublee studio         18m      780       Paris   /  30 août 2019   
113      meublee chambre          9m      400       Paris   /  30 août 2019   
123          appartement         68m     2095       Paris   /  30 août 2019   
149          appartement        105m     2900       Paris   /  30 août 2019   
185  meublee appartement         50m     2390       Paris   /  29 août 2019   
197       meublee studio         14m      850       Paris   /  29 août 2019   

                                           Description  
1    Daumesnil. Haussmannien, au calme, 3 pièces, 5...  
6    Meublé 3 pièces + dressing de 71 m² au 29ème é...  
11   Daumesnil. Haussmannien, au calme, 3 pièces, 5...  
16   Meublé 3 pièces + dressing de 71 m² au 29ème é...  
28   Bail meublé code civil (société ou hab.seconda...  
42   Bastille, près Place, Opéra.\r\nDans rue calme...  
64   Rue Raffet. 2 pièces meublé, 38 m² : entrée, s...  
82   Métro Duroc/Montparnasse.\r\nStudette atypique...  
90   2 pièces meublé 25 m², vue sur les toits de Pa...  
92   Métro Exelmans (lignes 9 et 10).\n\r\nBel imme...  
96   Boulevard Pereire, quartier Ternes bon côté du...  
99   Boulevard Voltaire. Dans bel immeuble Haussman...  
104  Rue Saint Dominique, métro La Tour Maubourg ou...  
105  Appartement meublé de 2 pièces situé au 61, ru...  
111  Panthéon. Studio meublé 18 m², au 6e sans asce...  
113  Métro Malesherbes ou Wagram. Boulevard Péreire...  
123  Proche Bastille. Appartement de 66 m² excellen...  
149  Loue appartement 105 m².\r\nRez-de-chaussée, t...  
185  Superbe 2 pièces 50 m². Rue St-Paul. Ascenseur...  
197  Au cœur du marché de Levis (46 rue de Levis). ...

In [12]:
pap_data[pap_data['Surface m²'] =='18m']
#checking if Paris only apparts are double counted in the Paris arrond rows

Type_appart Surface m²  Price € Localisation        Date Post  \
24           studio         18m      670    Paris 10E  /  30 août 2019   
39   meublee studio         18m      690    Paris 20E  /  30 août 2019   
43   meublee studio         18m      807    Paris 13E  /  30 août 2019   
47           studio         18m      730    Paris 13E  /  30 août 2019   
50   meublee studio         18m      750     Paris 9E  /  30 août 2019   
54   meublee studio         18m     1149     Paris 2E  /  30 août 2019   
77   meublee studio         18m      890     Paris 9E  /  30 août 2019   
104  meublee studio         18m      700       Paris   /  30 août 2019   
111  meublee studio         18m      780       Paris   /  30 août 2019   
173  meublee studio         18m      850     Paris 4E  /  30 août 2019   
179  meublee studio         18m     1080    Paris 16E  /  30 août 2019   

                                           Description  
24   Métro Louis Blanc.\n\r\nBeau studio loué vide ...  
39   Proximité Cours de Vincennes.\r\nStudio meublé...  
43   Studio meublé et équipé au 6e et dernier étage...  
47   Gobelins.  Studio 18 m², 1er étage ouvrant sur...  
50   Studio meublé 18 m², mansardé, refait neuf, 3 ...  
54   Montorgueil. Quartier Piétonnier.\r\nMétro: Se...  
77   Métro Cadet, quartier commerçant.\n\r\nStudio ...  
104  Rue Saint Dominique, métro La Tour Maubourg ou...  
111  Panthéon. Studio meublé 18 m², au 6e sans asce...  
173  Merci de me contacter de préférence par mail :...  
179  Studio meublé tout confort dans bel immeuble a...

In [13]:
pap_data.dtypes

Type_appart     object
Surface m²      object
Price €          int64
Localisation    object
Date Post       object
Description     object
dtype: object

In [14]:
pap_data['Surface m²'] = pap_data['Surface m²'].str.rstrip('m')

In [15]:
pap_data['Surface m²']=pap_data['Surface m²'].astype(int)

In [16]:
pap_data['Date Post']=pap_data['Date Post'].str.lstrip('/ ')

In [17]:
pap_data.head()

Type_appart  Surface m²  Price € Localisation          Date Post  \
0  meublee appartement           42     1280    Paris 14E  02 septembre 2019   
1  meublee appartement           53     1533       Paris        30 août 2019   
2  meublee appartement           38     1400    Paris 13E       30 août 2019   
3  meublee appartement           33     1130    Paris 18E       30 août 2019   
4  meublee appartement           75     2350     Paris 9E       30 août 2019   

                                         Description  
0  Métro Alésia, proche transports et bus.\n\r\nD...  
1  Daumesnil. Haussmannien, au calme, 3 pièces, 5...  
2  Métro place d'Italie ou Tolbiac.\r\n2 pièces m...  
3  2 pièces de 32,5m2 en parfait état entre Lamar...  
4  Appartement meublé 75,5 m², entièrement rénové...

In [18]:
pap_data['Meublee']= [i[0] for i in pap_data.Type_appart.str.split()]

In [19]:
import numpy as np

In [21]:
pap_data['Meublee'] = np.where(pap_data['Meublee']=='meublee', 1, 0)

In [22]:
pap_data.Type_appart=pap_data.Type_appart.str.strip('meublee').str.strip(' ')

In [23]:
pap_data.head()

Type_appart  Surface m²  Price € Localisation          Date Post  \
0  appartement          42     1280    Paris 14E  02 septembre 2019   
1  appartement          53     1533       Paris        30 août 2019   
2  appartement          38     1400    Paris 13E       30 août 2019   
3  appartement          33     1130    Paris 18E       30 août 2019   
4  appartement          75     2350     Paris 9E       30 août 2019   

                                         Description  Meublee  
0  Métro Alésia, proche transports et bus.\n\r\nD...        1  
1  Daumesnil. Haussmannien, au calme, 3 pièces, 5...        1  
2  Métro place d'Italie ou Tolbiac.\r\n2 pièces m...        1  
3  2 pièces de 32,5m2 en parfait état entre Lamar...        1  
4  Appartement meublé 75,5 m², entièrement rénové...        1

In [24]:
pap_data['Type_appart'].value_counts()

appartement    103
studio          89
chambre          8
Name: Type_appart, dtype: int64

In [25]:
pap_data.dtypes

Type_appart     object
Surface m²       int32
Price €          int64
Localisation    object
Date Post       object
Description     object
Meublee          int32
dtype: object

In [29]:
pap_data.Localisation.value_counts()

Paris 16E    21
Paris        20
Paris 15E    17
Paris 17E    15
Paris 18E    15
Paris 5E     14
Paris 11E    13
Paris 13E    10
Paris 19E    10
Paris 14E     9
Paris 6E      9
Paris 3E      7
Paris 12E     6
Paris 2E      6
Paris 9E      5
Paris 20E     5
Paris 10E     5
Paris 8E      4
Paris 7E      3
Paris 1Er     3
Paris 4E      2
Paris 11      1
Name: Localisation, dtype: int64

In [44]:
pap_data2=pap_data.rename(columns={'Localisation':'Luxury_location'})

In [53]:
pap_data2.Luxury_location = np.where(pap_data2.Luxury_location.isin(['Paris 1Er','Paris 2E','Paris 3E','Paris 5E','Paris 6E','Paris 7E','Paris 11','Paris 11E','Paris 4E','Paris 8E','Paris 9E','Paris 12E']),1,0)

In [55]:
pap_data2.Luxury_location.value_counts()

0    127
1     73
Name: Luxury_location, dtype: int64

In [27]:
pap_data['Code_postal']=pap_data['Localisation'].str.strip('Paris').str.strip(' ')
pap_data.Code_postal=pap_data.Code_postal.str.strip('E')

dict_arrond={'16':75016,'15':75015, '18':75018,'17':75017,'5':75005,'11':75011,'19':75019,'6':75006,'13':75013,'14':75014,'3':75003,'12':75012, '2':75002,'20':75020,'9':75009,'10':75010,'8':75008,'7':75007,'1':75001,'4':75004,'':75000}
pap_data.Code_postal=pap_data.Code_postal.map(dict_arrond)
pap_data.Code_postal

0      75014
1      75000
2      75013
3      75018
4      75009
5      75011
6      75000
7      75017
8      75002
9      75010
10     75014
11     75000
12     75013
13     75018
14     75009
15     75011
16     75000
17     75017
18     75002
19     75010
20     75010
21     75016
22     75003
23     75005
24     75010
25     75013
26     75016
27     75005
28     75000
29     75015
       ...  
170    75015
171    75017
172    75003
173    75004
174    75014
175    75013
176    75016
177    75006
178    75011
179    75016
180    75011
181    75003
182    75005
183    75017
184    75002
185    75000
186    75016
187    75017
188    75018
189    75011
190    75006
191    75014
192    75018
193    75019
194    75018
195    75020
196    75005
197    75000
198    75006
199    75015
Name: Code_postal, Length: 200, dtype: int64

In [56]:
pap_data2.dtypes

Type_appart        object
Surface m²          int32
Price €             int64
Luxury_location     int32
Date Post          object
Description        object
Meublee             int32
Code_postal         int64
dtype: object

In [57]:
pap_data2.head()

Type_appart  Surface m²  Price €  Luxury_location          Date Post  \
0  appartement          42     1280                0  02 septembre 2019   
1  appartement          53     1533                0       30 août 2019   
2  appartement          38     1400                0       30 août 2019   
3  appartement          33     1130                0       30 août 2019   
4  appartement          75     2350                1       30 août 2019   

                                         Description  Meublee  Code_postal  
0  Métro Alésia, proche transports et bus.\n\r\nD...        1        75014  
1  Daumesnil. Haussmannien, au calme, 3 pièces, 5...        1        75000  
2  Métro place d'Italie ou Tolbiac.\r\n2 pièces m...        1        75013  
3  2 pièces de 32,5m2 en parfait état entre Lamar...        1        75018  
4  Appartement meublé 75,5 m², entièrement rénové...        1        75009

In [71]:
pap_data3=pd.get_dummies(data=pap_data2, columns=['Type_appart','Code_postal'], drop_first=True)

In [202]:
pap_data3.columns

Index(['Surface m²', 'Price €', 'Luxury_location', 'Meublee',
       'Type_appart_chambre', 'Type_appart_studio', 'Code_postal_75001',
       'Code_postal_75002', 'Code_postal_75003', 'Code_postal_75004',
       'Code_postal_75005', 'Code_postal_75006', 'Code_postal_75007',
       'Code_postal_75008', 'Code_postal_75009', 'Code_postal_75010',
       'Code_postal_75011', 'Code_postal_75012', 'Code_postal_75013',
       'Code_postal_75014', 'Code_postal_75015', 'Code_postal_75016',
       'Code_postal_75017', 'Code_postal_75018', 'Code_postal_75019',
       'Code_postal_75020', 'Sold'],
      dtype='object')

In [74]:
pap_data3.to_csv('pap_data3_0830.csv', index=False)

In [138]:
pap_data3=pd.read_csv('pap_data3_0830.csv')

# Get PAP data on Sept 6th and check if flats to rent on Aug 30th have been rented

In [76]:
pap_data4=pd.read_csv('pap_location_data_0906.csv')
pap_data4.head()

Unnamed: 0           Type_appart Surface m²  Price € Localisation  \
0           0  meublee appartement         47m     1750     Paris 3E   
1           1  meublee appartement        130m     4800     Paris 6E   
2           2  meublee appartement         44m     1870     Paris 8E   
3           3       meublee studio         43m     1500    Paris 13E   
4           4          appartement         42m     1700    Paris 14E   

              Date Post                                        Description  
0  /  06 septembre 2019  Appartement de 2 pièces - meublé  - 47 m² - Pa...  
1  /  06 septembre 2019  Très bel appartement haussmannien avec moulure...  
2  /  06 septembre 2019  2 pièces au 8 rue Balzac, 3ème étage (2 ascens...  
3  /  06 septembre 2019  À partir du 15 novembre, Place d'Italie. Grand...  
4  /  06 septembre 2019  A quelques pas de Montparnasse !!!\r\nMétro Ed...

In [139]:
def check(x,y = pap_data4.Description) : 
    a = 1
    for i in y :
        if x == i :
            a = 0
    return a 
pap_data3['description_match'] = pap_data3.Description.map(lambda x : check(x))

In [140]:
pap_data3.description_match.head()

0    1
1    1
2    0
3    1
4    0
Name: description_match, dtype: int64

In [141]:
pap_data3=pap_data3.rename(columns={'description_match':'Sold'})

In [143]:
pap_data3.drop(['Date Post','Description'], axis=1, inplace=True)

In [204]:
pap_data3.head()

Surface m²   Price €  Luxury_location  Meublee  Type_appart_chambre  \
0    0.405660  0.275000                0        1                    0   
1    0.509434  0.354063                0        1                    0   
2    0.367925  0.312500                0        1                    0   
3    0.320755  0.228125                0        1                    0   
4    0.716981  0.609375                1        1                    0   

   Type_appart_studio  Code_postal_75001  Code_postal_75002  \
0                   0                  0                  0   
1                   0                  0                  0   
2                   0                  0                  0   
3                   0                  0                  0   
4                   0                  0                  0   

   Code_postal_75003  Code_postal_75004  ...  Code_postal_75012  \
0                  0                  0  ...                  0   
1                  0                  0  ...                  0   
2                  0                  0  ...                  0   
3                  0                  0  ...                  0   
4                  0                  0  ...                  0   

   Code_postal_75013  Code_postal_75014  Code_postal_75015  Code_postal_75016  \
0                  0                  1                  0                  0   
1                  0                  0                  0                  0   
2                  1                  0                  0                  0   
3                  0                  0                  0                  0   
4                  0                  0                  0                  0   

   Code_postal_75017  Code_postal_75018  Code_postal_75019  Code_postal_75020  \
0                  0                  0                  0                  0   
1                  0                  0                  0                  0   
2                  0                  0                  0                  0   
3                  0                  1                  0                  0   
4                  0                  0                  0                  0   

   Sold  
0     1  
1     1  
2     0  
3     1  
4     0  

[5 rows x 27 columns]

## Standardization

In [147]:
pap_data3.iloc[:,0:2]=pap_data3.iloc[:,0:2].apply(lambda x: (x-x.min())/(x.max()-x.min()))

# Build logistic regression model on Aug 30th data

In [189]:
import statsmodels.api as sm

In [205]:
logit=sm.Logit(pap_data3.Sold, pap_data3.drop('Sold', axis=1))
#use all data 

In [208]:
result=logit.fit()

Optimization terminated successfully.
         Current function value: 0.564598
         Iterations 7


In [209]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      175
Method:                           MLE   Df Model:                           24
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                  0.1280
Time:                        10:58:21   Log-Likelihood:                -112.92
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                    0.1012
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Surface m²              2.6562      2.269      1.171      0.242      -1.791       7.103
Price €                -5.4601      2.663     -2.051      0.040     -10.679      -0.241
Luxury_location         0.3607   2.85e+07   1.26e-08      1.000   -5.59e+07    5.59e+07
Meublee                 0.7023      0.365      1.924      0.054      -0.013       1.418
Type_appart_chambre    -0.7572      0.868     -0.873      0.383      -2.458       0.943
Type_appart_studio      0.4104      0.417      0.985      0.325      -0.406       1.227
Code_postal_75001      -1.0341   2.85e+07  -3.63e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75002       1.4263   2.85e+07      5e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75003       0.4181   2.85e+07   1.47e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75004      -0.0595   2.85e+07  -2.09e-09      1.000   -5.59e+07    5.59e+07
Code_postal_75005      -0.2971   2.85e+07  -1.04e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75006       0.2950   2.85e+07   1.03e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75007      -0.6827   2.85e+07  -2.39e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75008      -1.5275   2.85e+07  -5.36e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75009      -0.7976   2.85e+07   -2.8e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75010       0.0679      1.036      0.066      0.948      -1.962       2.098
Code_postal_75011       2.3983   2.85e+07   8.41e-08      1.000   -5.59e+07    5.59e+07
Code_postal_75012       0.2216   2.85e+07   7.77e-09      1.000   -5.59e+07    5.59e+07
Code_postal_75013       0.0057      0.782      0.007      0.994      -1.526       1.538
Code_postal_75014       0.6230      0.885      0.704      0.482      -1.112       2.358
Code_postal_75015       0.7371      0.654      1.127      0.260      -0.545       2.019
Code_postal_75016       0.3563      0.648      0.550      0.583      -0.914       1.627
Code_postal_75017       0.3599      0.691      0.521      0.603      -0.995       1.714
Code_postal_75018       1.1592      0.740      1.566      0.117      -0.291       2.610
Code_postal_75019       1.8346      0.934      1.964      0.050       0.004       3.666
Code_postal_75020       0.8771      1.216      0.721      0.471      -1.506       3.260
=======================================================================================
"""

In [213]:
sm.Logit(pap_data3.Sold, pap_data3.drop(['Sold','Code_postal_75001','Code_postal_75002','Code_postal_75003'], axis=1)).fit().summary()
#drop inrelevant variables (that have no relationship with the target), where p value > 0,05%

Optimization terminated successfully.
         Current function value: 0.569842
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      177
Method:                           MLE   Df Model:                           22
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                  0.1199
Time:                        11:15:19   Log-Likelihood:                -113.97
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                   0.09526
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Surface m²              2.4697      2.147      1.150      0.250      -1.739       6.678
Price €                -5.1270      2.486     -2.062      0.039     -10.000      -0.254
Luxury_location         0.9107      0.729      1.249      0.212      -0.519       2.340
Meublee                 0.6645      0.362      1.834      0.067      -0.046       1.375
Type_appart_chambre    -0.7377      0.864     -0.854      0.393      -2.431       0.955
Type_appart_studio      0.3695      0.410      0.900      0.368      -0.435       1.174
Code_postal_75004      -0.6134      1.713     -0.358      0.720      -3.970       2.744
Code_postal_75005      -0.8198      0.816     -1.004      0.315      -2.420       0.780
Code_postal_75006      -0.2669      0.898     -0.297      0.766      -2.027       1.493
Code_postal_75007      -1.2427      1.355     -0.917      0.359      -3.898       1.412
Code_postal_75008      -2.0574      1.355     -1.519      0.129      -4.713       0.598
Code_postal_75009      -1.3272      1.147     -1.157      0.247      -3.575       0.920
Code_postal_75010       0.1061      1.032      0.103      0.918      -1.916       2.128
Code_postal_75011       1.8308      1.217      1.505      0.132      -0.554       4.215
Code_postal_75012      -0.3515      1.023     -0.344      0.731      -2.357       1.653
Code_postal_75013       0.0399      0.777      0.051      0.959      -1.482       1.562
Code_postal_75014       0.6445      0.876      0.736      0.462      -1.072       2.361
Code_postal_75015       0.7378      0.650      1.135      0.256      -0.536       2.012
Code_postal_75016       0.3774      0.642      0.588      0.557      -0.881       1.635
Code_postal_75017       0.3759      0.686      0.548      0.584      -0.969       1.721
Code_postal_75018       1.1801      0.738      1.600      0.110      -0.266       2.626
Code_postal_75019       1.8066      0.926      1.951      0.051      -0.008       3.621
Code_postal_75020       0.9159      1.209      0.758      0.449      -1.453       3.285
=======================================================================================
"""

In [214]:
sm.Logit(pap_data3.Sold, pap_data3.drop(['Sold','Code_postal_75001','Code_postal_75002','Code_postal_75003','Code_postal_75013','Code_postal_75010','Code_postal_75012'], axis=1)).fit().summary()

Optimization terminated successfully.
         Current function value: 0.570167
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      180
Method:                           MLE   Df Model:                           19
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                  0.1194
Time:                        11:17:11   Log-Likelihood:                -114.03
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                   0.04128
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Surface m²              2.4439      2.080      1.175      0.240      -1.633       6.521
Price €                -5.0457      2.460     -2.051      0.040      -9.866      -0.225
Luxury_location         0.7741      0.601      1.289      0.197      -0.403       1.951
Meublee                 0.6669      0.356      1.875      0.061      -0.030       1.364
Type_appart_chambre    -0.7394      0.833     -0.888      0.375      -2.372       0.893
Type_appart_studio      0.3938      0.380      1.036      0.300      -0.351       1.138
Code_postal_75004      -0.5159      1.682     -0.307      0.759      -3.813       2.781
Code_postal_75005      -0.7140      0.749     -0.954      0.340      -2.182       0.754
Code_postal_75006      -0.1636      0.845     -0.194      0.847      -1.820       1.493
Code_postal_75007      -1.1361      1.317     -0.863      0.388      -3.718       1.445
Code_postal_75008      -1.9461      1.314     -1.480      0.139      -4.522       0.630
Code_postal_75009      -1.2186      1.093     -1.115      0.265      -3.360       0.923
Code_postal_75011       1.9411      1.171      1.658      0.097      -0.353       4.235
Code_postal_75014       0.6154      0.843      0.730      0.465      -1.037       2.268
Code_postal_75015       0.7129      0.615      1.159      0.247      -0.493       1.919
Code_postal_75016       0.3468      0.592      0.586      0.558      -0.813       1.507
Code_postal_75017       0.3505      0.643      0.545      0.586      -0.911       1.612
Code_postal_75018       1.1595      0.711      1.631      0.103      -0.234       2.553
Code_postal_75019       1.7827      0.904      1.971      0.049       0.010       3.555
Code_postal_75020       0.8927      1.193      0.748      0.454      -1.446       3.231
=======================================================================================
"""

In [215]:
sm.Logit(pap_data3.Sold, pap_data3.drop(['Sold','Code_postal_75001','Code_postal_75002','Code_postal_75003','Code_postal_75013','Code_postal_75010','Code_postal_75012','Code_postal_75016','Code_postal_75017','Code_postal_75014','Code_postal_75004','Code_postal_75006'], axis=1)).fit().summary()

Optimization terminated successfully.
         Current function value: 0.572577
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      185
Method:                           MLE   Df Model:                           14
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                  0.1156
Time:                        11:20:07   Log-Likelihood:                -114.52
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                  0.007759
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Surface m²              2.6297      2.101      1.252      0.211      -1.488       6.747
Price €                -4.8835      2.478     -1.971      0.049      -9.740      -0.027
Luxury_location         0.4968      0.474      1.049      0.294      -0.432       1.425
Meublee                 0.6938      0.352      1.974      0.048       0.005       1.383
Type_appart_chambre    -0.5828      0.797     -0.731      0.465      -2.145       0.979
Type_appart_studio      0.5178      0.342      1.513      0.130      -0.153       1.189
Code_postal_75005      -0.6428      0.692     -0.929      0.353      -1.999       0.713
Code_postal_75007      -1.0546      1.287     -0.819      0.413      -3.577       1.468
Code_postal_75008      -1.8410      1.283     -1.434      0.151      -4.356       0.674
Code_postal_75009      -1.1721      1.055     -1.111      0.266      -3.239       0.895
Code_postal_75011       2.0243      1.140      1.776      0.076      -0.209       4.258
Code_postal_75015       0.5333      0.577      0.924      0.356      -0.598       1.665
Code_postal_75018       1.0030      0.689      1.457      0.145      -0.347       2.353
Code_postal_75019       1.6093      0.876      1.838      0.066      -0.107       3.326
Code_postal_75020       0.7365      1.185      0.622      0.534      -1.585       3.058
=======================================================================================
"""

In [217]:
sm.Logit(pap_data3.Sold, pap_data3.drop(['Sold','Code_postal_75001','Code_postal_75002','Code_postal_75003','Code_postal_75013','Code_postal_75010','Code_postal_75012','Code_postal_75016','Code_postal_75017','Code_postal_75014','Code_postal_75004','Code_postal_75006','Code_postal_75020','Code_postal_75007','Code_postal_75005','Code_postal_75015'], axis=1)).fit().summary()

Optimization terminated successfully.
         Current function value: 0.579031
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      189
Method:                           MLE   Df Model:                           10
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                  0.1057
Time:                        11:22:26   Log-Likelihood:                -115.81
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                  0.002279
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Surface m²              2.7118      2.060      1.316      0.188      -1.326       6.750
Price €                -4.6349      2.434     -1.904      0.057      -9.406       0.136
Luxury_location         0.1254      0.386      0.325      0.745      -0.630       0.881
Meublee                 0.6821      0.346      1.970      0.049       0.003       1.361
Type_appart_chambre    -0.5780      0.793     -0.729      0.466      -2.132       0.976
Type_appart_studio      0.5411      0.335      1.617      0.106      -0.115       1.197
Code_postal_75008      -1.5590      1.250     -1.247      0.212      -4.009       0.891
Code_postal_75009      -0.9064      1.015     -0.893      0.372      -2.896       1.083
Code_postal_75011       2.2584      1.113      2.029      0.042       0.077       4.440
Code_postal_75018       0.9192      0.683      1.345      0.178      -0.420       2.258
Code_postal_75019       1.4764      0.862      1.712      0.087      -0.214       3.166
=======================================================================================
"""

In [218]:
sm.Logit(pap_data3.Sold, pap_data3.drop(['Sold','Code_postal_75001','Code_postal_75002','Code_postal_75003','Code_postal_75013','Code_postal_75010','Code_postal_75012','Code_postal_75016','Code_postal_75017','Code_postal_75014','Code_postal_75004','Code_postal_75006','Code_postal_75020','Code_postal_75007','Code_postal_75005','Code_postal_75015','Code_postal_75009','Code_postal_75008','Code_postal_75018','Code_postal_75019'], axis=1)).fit().summary()

Optimization terminated successfully.
         Current function value: 0.598072
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      193
Method:                           MLE   Df Model:                            6
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                 0.07626
Time:                        11:24:51   Log-Likelihood:                -119.61
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                  0.003068
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Surface m²              3.2428      1.986      1.633      0.102      -0.649       7.135
Price €                -4.6449      2.371     -1.959      0.050      -9.291       0.001
Luxury_location        -0.2093      0.356     -0.588      0.557      -0.907       0.489
Meublee                 0.5859      0.331      1.767      0.077      -0.064       1.236
Type_appart_chambre    -0.5306      0.779     -0.681      0.496      -2.057       0.996
Type_appart_studio      0.6489      0.327      1.986      0.047       0.009       1.289
Code_postal_75011       2.3862      1.099      2.172      0.030       0.233       4.540
=======================================================================================
"""

In [219]:
sm.Logit(pap_data3.Sold, pap_data3.drop(['Sold','Code_postal_75001','Code_postal_75002','Code_postal_75003','Code_postal_75013','Code_postal_75010','Code_postal_75012','Code_postal_75016','Code_postal_75017','Code_postal_75014','Code_postal_75004','Code_postal_75006','Code_postal_75020','Code_postal_75007','Code_postal_75005','Code_postal_75015','Code_postal_75009','Code_postal_75008','Code_postal_75018','Code_postal_75019','Luxury_location'], axis=1)).fit().summary()

Optimization terminated successfully.
         Current function value: 0.598932
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      194
Method:                           MLE   Df Model:                            5
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                 0.07493
Time:                        11:28:56   Log-Likelihood:                -119.79
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                  0.001615
=======================================================================================
                          coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------
Surface m²              3.5199      1.948      1.806      0.071      -0.299       7.339
Price €                -5.1014      2.266     -2.251      0.024      -9.542      -0.660
Meublee                 0.5662      0.330      1.714      0.086      -0.081       1.214
Type_appart_chambre    -0.5863      0.771     -0.760      0.447      -2.098       0.926
Type_appart_studio      0.6232      0.324      1.925      0.054      -0.011       1.258
Code_postal_75011       2.2422      1.072      2.091      0.036       0.141       4.343
=======================================================================================
"""

In [239]:
from statsmodels.tools.tools import add_constant
final_model=sm.Logit(pap_data3.Sold, pap_data3.drop(['Sold','Code_postal_75001','Code_postal_75002','Code_postal_75003','Code_postal_75013','Code_postal_75010','Code_postal_75012','Code_postal_75016','Code_postal_75017','Code_postal_75014','Code_postal_75004','Code_postal_75006','Code_postal_75020','Code_postal_75007','Code_postal_75005','Code_postal_75015','Code_postal_75009','Code_postal_75008','Code_postal_75018','Code_postal_75019','Code_postal_75011','Luxury_location','Type_appart_chambre'], axis=1)).fit(method='lbfgs')
final_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                   Sold   No. Observations:                  200
Model:                          Logit   Df Residuals:                      196
Method:                           MLE   Df Model:                            3
Date:                Fri, 13 Sep 2019   Pseudo R-squ.:                 0.04378
Time:                        11:54:44   Log-Likelihood:                -123.82
converged:                       True   LL-Null:                       -129.49
                                        LLR p-value:                   0.01004
======================================================================================
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
Surface m²             3.9007      1.885      2.069      0.039       0.206       7.595
Price €               -5.1703      2.180     -2.372      0.018      -9.443      -0.898
Meublee                0.4936      0.302      1.635      0.102      -0.098       1.085
Type_appart_studio     0.6750      0.313      2.154      0.031       0.061       1.289
======================================================================================
"""

In [335]:
pap_data3.shape

(200, 27)

# After building the model on August 30th data we will test on data from Sept 6th

In [312]:
test=pd.read_csv('pap_location_data_0906.csv')
test.head()

Unnamed: 0           Type_appart Surface m²  Price € Localisation  \
0           0  meublee appartement         47m     1750     Paris 3E   
1           1  meublee appartement        130m     4800     Paris 6E   
2           2  meublee appartement         44m     1870     Paris 8E   
3           3       meublee studio         43m     1500    Paris 13E   
4           4          appartement         42m     1700    Paris 14E   

              Date Post                                        Description  
0  /  06 septembre 2019  Appartement de 2 pièces - meublé  - 47 m² - Pa...  
1  /  06 septembre 2019  Très bel appartement haussmannien avec moulure...  
2  /  06 septembre 2019  2 pièces au 8 rue Balzac, 3ème étage (2 ascens...  
3  /  06 septembre 2019  À partir du 15 novembre, Place d'Italie. Grand...  
4  /  06 septembre 2019  A quelques pas de Montparnasse !!!\r\nMétro Ed...

In [313]:
test.drop(columns=['Unnamed: 0', 'Date Post', 'Description','Localisation'], inplace=True)

In [314]:
test['Surface m²'] = test['Surface m²'].str.rstrip('m').astype(int)

In [318]:
test['Meublee']= [i[0] for i in test.Type_appart.str.split()]
test['Meublee'] = np.where(test['Meublee']=='meublee', 1, 0)
test.Type_appart=test.Type_appart.str.strip('meublee').str.strip(' ')
test = test.replace('studio Paris', 'studio' )

In [327]:
test.Type_appart.value_counts()

appartement    108
studio          86
chambre          6
Name: Type_appart, dtype: int64

In [328]:
test=pd.get_dummies(data=test, columns=['Type_appart'], drop_first=True)

In [329]:
test.iloc[:,0:2]=test.iloc[:,0:2].apply(lambda x: (x-x.min())/(x.max()-x.min()))

In [331]:
test.drop(['Type_appart_chambre'], axis=1, inplace=True)

In [332]:
test.head()

Surface m²   Price €  Meublee  Type_appart_studio
0    0.366412  0.364716        1                   0
1    1.000000  1.000000        1                   0
2    0.343511  0.389710        1                   0
3    0.335878  0.312643        1                   1
4    0.328244  0.354301        0                   0

In [333]:
test.shape

(200, 4)

2. Get predictions Sold on Sept 6th data

In [336]:
predictions=round(final_model.predict(test))

In [353]:
predictions.value_counts()

1.0    149
0.0     51
dtype: int64

Predictive model results are : 149 appartments have been sold/rented and 51 appartments have not been sold/rented

# Now let's compare the predictive model result with the real data scrapped on the site on Sept 13th

In [346]:
pap_data1309=pd.read_csv('PAP_data1309_forcomp0609.csv')

In [347]:
pap_data1309.head()

Unnamed: 0           Type_appart Surface m²  Price € Localisation  \
0           0          appartement         79m     1.99    Paris 17E   
1           1       meublee studio         33m     1.32    Paris 14E   
2           2          appartement         57m     1.52    Paris 16E   
3           3  meublee appartement         76m     2.90    Paris 16E   
4           4  meublee appartement         33m     1.59     Paris 5E   

              Date Post                                        Description  
0  /  13 septembre 2019  Location vide.\n\r\nAppartement 3 pièces de 79...  
1  /  13 septembre 2019  \n\t\t\t\t\tGrand studio moderne entre Montpar...  
2  /  13 septembre 2019  Appartement 2 pièces de 57 m², quartier Exelma...  
3  /  13 septembre 2019  Très agréable appartement dans un bel immeuble...  
4  /  13 septembre 2019  Quartier Latin / Les Gobelins  - 33m2.\n\r\nMa...

In [348]:
pap_data0609=pd.read_csv('pap_location_data_0906.csv')
pap_data0609.head()

Unnamed: 0           Type_appart Surface m²  Price € Localisation  \
0           0  meublee appartement         47m     1750     Paris 3E   
1           1  meublee appartement        130m     4800     Paris 6E   
2           2  meublee appartement         44m     1870     Paris 8E   
3           3       meublee studio         43m     1500    Paris 13E   
4           4          appartement         42m     1700    Paris 14E   

              Date Post                                        Description  
0  /  06 septembre 2019  Appartement de 2 pièces - meublé  - 47 m² - Pa...  
1  /  06 septembre 2019  Très bel appartement haussmannien avec moulure...  
2  /  06 septembre 2019  2 pièces au 8 rue Balzac, 3ème étage (2 ascens...  
3  /  06 septembre 2019  À partir du 15 novembre, Place d'Italie. Grand...  
4  /  06 septembre 2019  A quelques pas de Montparnasse !!!\r\nMétro Ed...

In [350]:
def check(x,y = pap_data1309.Description) : 
    a = 1
    for i in y :
        if x == i :
            a = 0
    return a 
pap_data0609['Sold'] = pap_data0609.Description.map(lambda x : check(x))

In [352]:
pap_data0609.Sold.value_counts()

1    136
0     64
Name: Sold, dtype: int64

the real data match shows us that 136 appartments have been sold/rented and 64 appartments have not been sold/rented

# Predictive Model looks pretty good now let's build the Confusion Matrix

In [354]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(pap_data0609.Sold, predictions)

print(cnf_matrix)

print("Accuracy:",metrics.accuracy_score(pap_data0609.Sold, predictions))

print("Precision:",metrics.precision_score(pap_data0609.Sold, predictions))

print("Recall:",metrics.recall_score(pap_data0609.Sold, predictions))

[[ 18  46]
 [ 33 103]]
Accuracy: 0.605
Precision: 0.6912751677852349
Recall: 0.7573529411764706
